### ISSUE TITLE
Payload Executor ISA considerations

### ISSUE THREAD

**Issue by jedrzejboczar** [antmicro/litex-rowhammer-tester#6]
I created an initial version of the [Payload Executor module](https://github.com/antmicro/litex-rowhammer-tester/blob/9e35e5071629cfde7d6e78c1ac9c6810fb5286d0/gateware/payload_executor.py) along with [an example payload](https://github.com/antmicro/litex-rowhammer-tester/blob/9e35e5071629cfde7d6e78c1ac9c6810fb5286d0/scripts/execute_payload.py#L14-L34). 

The current version supports most of the required features. It implements all the needed DFI commands (ACT, PRE, REF, READ, ZQC). Each these instructions has a timeslice field defining the delay (in cycles) after the command and an address field. There is also a NOOP instruction that uses just the timeslice (timeslice field is wider because it reuses address bits) and a LOOP instruction which has count/jump fields. LOOP is a relative jump back by _jump_ instructions that gets executed _count_ times ([details](https://github.com/antmicro/litex-rowhammer-tester/blob/9e35e5071629cfde7d6e78c1ac9c6810fb5286d0/gateware/payload_executor.py#L253-L264)) and in this implementation has an execution time of 1.

The scratchpad memory is also implemented. It stores the data from subsequent READs at subsequent addresses. If an overflow happens it will be signalized in the status register. Both payload an scratchpad memories can be simply accessed from the main bus.

The current design of the ISA has one limitation which is that the Payload Executor works at the controller frequency and we have a 1:4 frequency ratio between the controller and the DRAM I/Os (so we have 4 DFI phases). Currently there is no way to specify the DFI phase and all the commands are executed on phase 0. This means that the the time resolution is 4 x tCK. In practice this may or may not be a problem, because it is often impossible to send commands so often due to DRAM module timing restrictions. It would be fairly easy to add an additional PHASE field to DFI instructions (reducing the width of the TIMESLICE field), which would allow to send the commands on any phase (with 1 x tCK resolution), but still at most 1 command per 4 x tCK.

As for the the bit widths of instruction fields, they are defined [here](https://github.com/antmicro/litex-rowhammer-tester/blob/9e35e5071629cfde7d6e78c1ac9c6810fb5286d0/gateware/payload_executor.py#L49-L69) and can be easily adjusted to suit our needs (especially LOOP count vs jump).

**Comment by yoongu** [comment]
Thanks for the update! If you're running DDR3-800, then your clock should be 400MHz, so is tCK = 2.5ns? 10ns isn't bad at all for our needs.

This looks mostly good. One nitpick: I think the opcode bit order should go RAS, CAS, WE (most-to-least significant) just like how they have it in the standard. Let us take a closer look and get back to you with any comments.

![image](https://user-images.githubusercontent.com/159680/97527526-eb0e7180-1981-11eb-88dd-303f8f4c8cd2.png)

**Comment by jedrzejboczar** [comment]
Yes, currently we have tCK = 2.5ns. I just modified the op codes to use RAS, CAS, WE order, the changes are [on this branch](https://github.com/antmicro/litex-rowhammer-tester/blob/jboc/dev/gateware/payload_executor.py).

**Comment by yoongu** [comment]
I whipped up a proto definition for the payload.

```
syntax = "proto2";

package dram.ddr3.proto;

enum Opcode {
  ACT = 4;  // RAS: 1, CAS: 0, WE: 0
  PRE = 5;  // RAS: 1, CAS: 0, WE: 1
  REF = 6;  // RAS: 1, CAS: 1, WE: 0
  NOP = 0;
  JMP = 7;
  MAX = 8;  // Unused
}

message Instr {
  // Refer to "DDR3 SDRAM Standard". JESD79-3F. July 2012. 
  message MemInstr {
    enum Bits {
      option allow_alias = true;
      OPCODE = 3;
      TIMESLICE = 8;
      RANK = 2;
      BANK = 3;
      ADDR = 16;
    }

    optional Opcode opcode = 1;
    optional uint32 timeslice = 2;
    optional uint32 rank = 3;
    optional uint32 bank = 4;
    optional uint32 addr = 5;
  }

  message NopInstr {
    enum Bits {
      OPCODE = 3;
      TIMESLICE = 29;
    }

    optional Opcode opcode = 1 [default = NOP];
    optional uint32 timeslice = 2;
  }

  message JmpInstr {
    enum Bits {
      OPCODE = 3;
      OFFSET = 14;
      COUNT = 15;
    }

    optional Opcode opcode = 1 [default = JMP];
    optional uint32 offset = 2;
    optional uint32 count = 3;
  }

  oneof instr {
    MemInstr mem = 1;
    NopInstr nop = 2;
    JmpInstr jmp = 3;
  }
}

// Represents DRAM timing parameters as positive integral multiples of the
// system clock: i.e., tick.
message Timing {
  // Unit: picoseconds.
  optional uint32 tick_ps = 1;
  // Unit: tick.
  optional uint32 ras = 2;
  optional uint32 rp = 3;
  optional uint32 rfc = 4;
}

message Payload {
  optional Timing timing = 1;
  repeated Instr instr = 2;
}
```

**Comment by yoongu** [comment]
And here's what a human-readable text proto would look like. We can handcraft these, validate them for timing, and run them.

```
timing: {
  tick_ps: 1000
  ras: 35
  rp: 15
  rfc: 100
}
instr: {
  mem: {
    opcode: ACT
    timeslice: 35
  }
}
instr: {
  mem: {
    opcode: PRE
    timeslice: 14
  }
}
instr: {
  jmp: {
    offset: 2
    count: 999999
  }
}
```

**Comment by yoongu** [comment]
I checked that this compiles and works correctly on a C++ shim that I'm bringing up. It should also work for Python. Let me know what you think.

**Comment by jedrzejboczar** [comment]
It looks good, should be quite easy to adjust field widths if we find that we need more bits for some fields, as e.g. the loop count value of `999999` in your example wouldn't fit in the instruction's 15-bit field.

**Comment by yoongu** [comment]
Agreed, having 14 bits for the offset seems too high. How about 9/20 for offset/count? That'll let us specify a hammer of 1M right out of the gate. Which is the order of magnitude that you measured in Arty. Let me send in another pull request later.

**Comment by jedrzejboczar** [comment]
@yoongu I merged the PR with proto definition, merged the Payload Executor to master and updated the field widths in code to match. Later we could probably load the field widths in Python from the proto file.

I also fixed the problem with scratchpad and had to make a change to the way payload is executed: https://github.com/antmicro/litex-rowhammer-tester/issues/7#issuecomment-721736948. Now all DFI commands are sent on phase 0 except READ command, which will be sent on `rdphase` which depends on the PHY and clock frequency. For the current setup it is phase 2, so READ commands are sent on `4n * tCK + 2`, where `n` is the instruction number.
